src/neuroconv/tools/nwb_helpers/_dataset_configuration.py

```diff
@@ -0,0 +1,204 @@
+from typing import Generator, Literal, Union
+
+import h5py
+import numpy as np
+import zarr
+from hdmf import Container
+from hdmf.data_utils import DataChunkIterator, DataIO, GenericDataChunkIterator
+from hdmf.utils import get_data_shape
+from hdmf_zarr import NWBZarrIO
+from pynwb import NWBHDF5IO, NWBFile, TimeSeries
+from pynwb.base import DynamicTable
+
+from ._models._base_models import DatasetInfo, DatasetIOConfiguration
+from ._models._hdf5_models import HDF5BackendConfiguration, HDF5DatasetIOConfiguration
+from ._models._zarr_models import ZarrBackendConfiguration, ZarrDatasetIOConfiguration
+from ..hdmf import SliceableDataChunkIterator
+
+BACKEND_TO_DATASET_CONFIGURATION = dict(hdf5=HDF5DatasetIOConfiguration, zarr=ZarrDatasetIOConfiguration)
+BACKEND_TO_CONFIGURATION = dict(hdf5=HDF5BackendConfiguration, zarr=ZarrBackendConfiguration)
+
+
+def _get_io_mode(io: Union[NWBHDF5IO, NWBZarrIO]) -> str:
+    """NWBHDF5IO and NWBZarrIO have different ways of storing the io mode (e.g. "r", "a", "w") they used on a path."""
+    if isinstance(io, NWBHDF5IO):
+        return io.mode
+    elif isinstance(io, NWBZarrIO):
+        return io._ZarrIO__mode
+
+
+def _is_dataset_written_to_file(
+    candidate_dataset: Union[h5py.Dataset, zarr.Array],
+    backend: Literal["hdf5", "zarr"],
+    existing_file: Union[h5py.File, zarr.Group, None],
+) -> bool:
+    """
+    Determine if the neurodata object is already written to the file on disk.
+
+    This object should then be skipped by the `get_io_datasets` function when working in append mode.
+    """
+    return (
+        isinstance(candidate_dataset, h5py.Dataset)  # If the source data is an HDF5 Dataset
+        and backend == "hdf5"
+        and candidate_dataset.file == existing_file  # If the source HDF5 Dataset is the appending NWBFile
+    ) or (
+        isinstance(candidate_dataset, zarr.Array)  # If the source data is an Zarr Array
+        and backend == "zarr"
+        and candidate_dataset.store == existing_file  # If the source Zarr 'file' is the appending NWBFile
+    )
+
+
+def _parse_location_in_memory_nwbfile(current_location: str, neurodata_object: Container) -> str:
+    parent = neurodata_object.parent
+    if isinstance(parent, NWBFile):
+        # Items in defined top-level places like acquisition, intervals, etc. do not act as 'containers'
+        # in that they do not set the `.parent` attribute; ask if object is in their in-memory dictionaries instead
+        for parent_field_name, parent_field_value in parent.fields.items():
+            if isinstance(parent_field_value, dict) and neurodata_object.name in parent_field_value:
+                return parent_field_name + "/" + neurodata_object.name + "/" + current_location
+        return neurodata_object.name + "/" + current_location
+    return _parse_location_in_memory_nwbfile(
+        current_location=neurodata_object.name + "/" + current_location, neurodata_object=parent
+    )
```

Comment:
Bear with me on this one as I was interested on building this at some point.  Would this be equivalent to this:

```python
def _parse_location_in_memory_nwbfile(current_location: str, neurodata_object: Container) -> str:
    parent = neurodata_object.parent
    
    is_top_level = isinstance(parent, NWBFile) 
    
    if not is_top_level:
        # Keep going up the tree until we find the top-level NWBFile
        updated_location = neurodata_object.name + "/" + current_location
        _parse_location_in_memory_nwbfile(current_location=updated_location, neurodata_object=parent)   
    else:
        # Items in defined top-level places like acquisition, intervals, etc. do not act as 'containers'
        # in that they do not set the `.parent` attribute; ask if object is in their in-memory dictionaries instead
        is_group = isinstance(parent, dict)
        if is_group:
            for parent_field_name, parent_field_value in parent.fields.items():
                if isinstance(parent_field_value, dict) and neurodata_object.name in parent_field_value:
                    return parent_field_name + "/" + neurodata_object.name + "/" + current_location
        else: # This is a dataset in the top-level NWBFile?
            return neurodata_object.name + "/" + current_location
```

Reply:
Best way to test would be to checkout a new branch off of this one and replace that method with your approach, see if the tests produce equivalent results

Reply:
All right, I will give it a try once we merge this : )